platform/android/src/snapshotter/map_snapshotter.cpp

```diff
@@ -0,0 +1,110 @@
+#include "map_snapshotter.hpp"
+
+#include <mbgl/renderer/renderer.hpp>
+#include <mbgl/style/style.hpp>
+#include <mbgl/util/shared_thread_pool.hpp>
+#include <mbgl/util/logging.hpp>
+#include <mbgl/actor/scheduler.hpp>
+
+#include "../attach_env.hpp"
+#include "../bitmap.hpp"
+
+namespace mbgl {
+namespace android {
+
+MapSnapshotter::MapSnapshotter(jni::JNIEnv& _env,
+                               jni::Object<MapSnapshotter> _obj,
+                               jni::Object<FileSource> jFileSource,
+                               jni::jfloat _pixelRatio,
+                               jni::jint width,
+                               jni::jint height,
+                               jni::String styleURL,
+                               jni::Object<LatLngBounds> region,
+                               jni::Object<CameraPosition> position,
+                               jni::String _programCacheDir)
+        : javaPeer(SeizeGenericWeakRef(_env, jni::Object<MapSnapshotter>(jni::NewWeakGlobalRef(_env, _obj.Get()).release())))
+        , pixelRatio(_pixelRatio)
+        , threadPool(sharedThreadPool()) {
+
+    // Get a reference to the JavaVM for callbacks
+    if (_env.GetJavaVM(&vm) < 0) {
+        _env.ExceptionDescribe();
+        return;
+    }
+
+    auto& fileSource = mbgl::android::FileSource::getDefaultFileSource(_env, jFileSource);
+    auto size = mbgl::Size { static_cast<uint32_t>(width), static_cast<uint32_t>(height) };
+    auto cameraOptions = position ? CameraPosition::getCameraOptions(_env, position) : CameraOptions();
+    optional<mbgl::LatLngBounds> bounds;
+    if (region) {
+        bounds = LatLngBounds::getLatLngBounds(_env, region);
+    }
+
+    // Create the core snapshotter
+    snapshotter = std::make_unique<mbgl::MapSnapshotter>(fileSource,
+                                                         *threadPool,
+                                                         jni::Make<std::string>(_env, styleURL),
+                                                         size,
+                                                         pixelRatio,
+                                                         cameraOptions,
+                                                         bounds,
+                                                         jni::Make<std::string>(_env, _programCacheDir));
+
+}
+
+MapSnapshotter::~MapSnapshotter() = default;
+
+void MapSnapshotter::start(JNIEnv&) {
+    MBGL_VERIFY_THREAD(tid);
+
+    snapshotCallback = std::make_unique<Actor<SnapshotCallback>>(*Scheduler::GetCurrent(), [this](std::exception_ptr err, PremultipliedImage image) {
+        MBGL_VERIFY_THREAD(tid);
+
+        if (err) {
+            try {
+                std::rethrow_exception(err);
```

Comment:
Thanks. I will add an additional callback. I refrained from doing so initially as there is not much you can do in these situations.